design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+pre-configured, and then capable of in operation management.
+
+#### Story 3
+As an admin, when creating a new machine, it is possible to automate configure
+LAG/mLAG.
+
+#### Story 4
+As an admin, when creating a cluster, it is possible to configure ACL settings
+in a switch for each bare metal server according to its functionality.
+The ACL setting:
+- Direction: ingress or egress
+- Protocol: SSH, HTTP, …all
+- Action: allow, deny
+- Port status: up, down (status of the switch port)
+
+## Design Details
+
+Resources and controllers scheduled to be added and modified in the current
+design:
+- Switch CRD:
+  - Switch CR represents a physical switch. It contains the system information
+    (vendor, OS, ...) and the access information of the switch.
+- SwitchPort CRD:
+  - SwitchPort CR represents one port in a switch. It contains the
+    configuration information of a normal port on the switch, the connection
+    information, the information of the switch to which it belongs, and a field
+    indicating whether the configuration is completed.
+- Switch controller:
+  - Monitors the above two resources, collects all the changes in all
+    SwitchPort CRs, configures the switch, and sets the field where the
+    SwitchPort configuration is completed.
+- Network CRD:
+  - Contains the network configuration of the cluster and the references
+    referring to which interfaces are currently using this network
+    configuration.
+- Network controller:
+  - Monitors Network CRs, when there is a new interface using the configuration
+    defined by a Network CR, changes the corresponding SwitchPort CR vlan
+    configuration.
+- Interface CRD:
+  - Interface CR represents a network interface, this can be a NIC in a bare
+    metal machine. It contains device information, LLDP information including
+    the connected switch’s mac address and the switch portID of the link, and
+    the reference referring to the device which this interface is connected to.
+- Interface controller:
+  - Monitors Interface CRs. When there is a new Interface CR, find the
+    corresponding SwitchPort CR according to the provided information and store
+    the reference of that SwitchPort CR in the Interface CR.
+- CAPM3:
+  - Gets the LLDP information of all the interfaces in the host. If there is no
+    corresponding Interface CR, create one.
+  - If a host has multiple interfaces, selects one for network configuration.
+  - Queries the configuration status of the SwitchPort CR to ensure that the
+    configuration is successful before provisioning.
+
+The main implementation process is described below:
+- After creating a BMH, the administrator puts an interface into the
+  provisioning network.
+
+  ![automatically configure 1](images/automatically_configure_1.png)
+
+- After an Interface is put into the provisioning network, the LLDP information
+  of other interfaces can be obtained and network configuration can be
+  performed.
```

Comment:
This section needs to be rewritten to cover what is doing each action (obtaining the LLDP information and performing the network configuration).